### ISSUE TITLE
Next generation buildout

### ISSUE THREAD

**Issue by sirex** [buildout/buildout#285]
Buildout is really awesome tool, but it starting to become outdated.

To fix buildout here is a list of things, that needs to be fixed:

1. Deprecate `zc.recipe.egg` and replace it with virtualenv and pip. Buildout should not create it's own `bin`, `eggs`, `parts` and etc. Instead it should use a virtualenv directory and install packages using `pip` tools.

2. Many recipes, like `z3c.recipe.mkdir` does not require virtualenv at all, so, virtualenv directory should be created only if needed.

3. Already existing virtualenv directory should be reused.

4. I like automatic version pinning using `versions.cfg` file, so this feature should be preserved.

5. Next generation buildout version should be backwards incompatible and started as new project taking only good parts from old `zc.buildout`. But it would be nice if existing recipes would be supported.

6. Buildout should be able to download specified virtualenv and pip version for creating new virtualenv.

7. Recipe API should be more simple, single function should be enough to define a recipe.

In other words, buildout should be recipe manager and wrapper around virtualenv and pip.

What I like in zc.buildout:

- Automatic version pinning in `versions.cfg`.

- `z3c.recipe.tag`.

- `djangorecipe`.

- `z3c.recipe.mkdir`.

- `hexagonit.recipe.download`.

- Script generation with parameters or environment variables set in generated script.

- Consistent state of packages and versions. Currently `pip freeze` just dumps everything you have in virtualenv, instead of packages and all its dependencies specified in `setup.py` install requires.

**Comment by reinout** [comment]
I don't agree with all the points, but you're right that some modernization is in order. Wheel support, for instance, which pip supports, starts to be a real problem.

An important mail to read is this one: https://mail.python.org/pipermail/distutils-sig/2014-December/025342.html (and preferably the whole thread by that name: https://mail.python.org/pipermail/distutils-sig/2014-December/thread.html#25317 ).

Some things to keep in mind:

- Pip doesn't have an API. If we want to use it, it'll be the horrible equivalent of `os.system('pip install something')`...

- https://github.com/pypa/packaging *is* a reusable library that does everything we would want to do. Provided it works well enough and provided it is ready for generic usage. I haven't checked it out yet. As mentioned in the abovementioned mailinglist thread: "it is worrying that Pip doesn't use it". As far as I know, no one that works on buildout has really checked it out yet.

- Setuptools, which we use behind the scenes, is being worked on again. There's a quiet hope that it'll start supporting wheels.

- Once you use Pip, you lose the ability to do some buildout-specific tricks. **Djangorecipe**, which you mention, uses buildout to add the django settings file environment variable to the scripts it generates, for instance. I don't know if that would be possible/easy with pip.

- Keep an eye on backwards compatibility. The changes you propose would, I think, boil down to a complete incompatible rewrite. Including rewrites of most of the recipes. It is hard to gain lots of users that way.

**Comment by sirex** [comment]
Well this is just a crazy idea, I just wanted to know how much support I could get from existing buildout developers.

> * Pip doesn't have an API. If we want to use it, it'll be the horrible equivalent of os.system('pip install something')...
>
> * https://github.com/pypa/packaging is a reusable library that does everything we would want to do. Provided it works well enough and provided it is ready for generic usage. I haven't checked it out yet. As mentioned in the abovementioned mailinglist thread: "it is worrying that Pip doesn't use it". As far as I know, no one that works on buildout has really checked it out yet.
> 
> * Setuptools, which we use behind the scenes, is being worked on again. There's a quiet hope that it'll start supporting wheels.

I have done some scripting with pip. There is no official supported API for pip, but unofficial API is quite usable. And in my opinion the whole official API think is just a written documentation, that can come later. So I don't see issues there.

> * Once you use Pip, you lose the ability to do some buildout-specific tricks. Djangorecipe, which you mention, uses buildout to add the django settings file environment variable to the scripts it generates, for instance. I don't know if that would be possible/easy with pip.

Script generation is really good feature that comes with buildout. I don't really know how buildout does it at the moment, but with Django it looks quite easy, you just need to generate small script with information provided by `console_script` entry point and add some code before executing entry point: https://github.com/django/django/blob/master/setup.py#L48

> * Keep an eye on backwards compatibility. The changes you propose would, I think, boil down to a complete incompatible rewrite. Including rewrites of most of the recipes. It is hard to gain lots of users that way.

Backwards compatibility would be really nice, but I see many people just stop using buildout at all. So in this case, I think new up to date buildout for new projects and users is much more important than providing backwards compatibility. Of course if it is possible to somehow easily implement backwards compatibility for existing recipes, that would be nice.

I think is easy to get a lot of users just by officially deprecating old buildout ant starting to support new one.

**Comment by jimfulton** [comment]
Some high-level comments.

1. Buildout really needs an API that it can use. Given the convergence of the rest of the Python community on pip, I'd prefer in the long run that buildout use the same code that pip does to fetch/install packages. I don't want to shell out to pip though.

2. I think jar files was something that Java got right.  I've done a fair bit of Scala development and having programs just define a class path as a list of jar paths works really well.  Of course, this is exactly how buildout uses eggs.  It pisses me off that eggs have gotten the bad rap they have. While there were some legitimate complaints, I think there was a lot of FUD, misinformation and generally unfortunate politics around the rejection of eggs. 

Nevertheless, it is what it is. Wheels could probably be used like or converted to eggs, but I'm not sure if anyone is willing to sign up for that effort.  The main advantage of eggs is that it lets you control versions on a script by script bases. I think it also makes it easier to change versions used. (I'm unsure what the current status of pip uninstalling things is.)

In practice, most projects use a single set of versions, so technically we don't need the benefits of eggs. (Though, again, I'm still unsure how well pip does uninstalling/upgrading.)  I think as a community, we have to ask ourselves whether the benefits of eggs outweigh the cost of maintaining our own packaging software, or depending of software that, increasingly, no one else uses. When I set out to write buildout, I certainly didn't want to write as much packaging code as I ultimately did. I wanted to just reuse easy_install/setuptools.

We've been able to get by for a long time with a wait-and-see attitude. Obviously, this won't work indefinitely.

I think the jar/eggs approach is elegant, but *I* am not willing to commit significant effort to do the work over the long term to keep making it work without support from the wider Python packaging community.  I'm not sure if someone else is, but I doubt it. :)

I think one of the first things we should do is have a conversation in the wider buildout community about whether people feel strongly that we need to retain the capability of having multiple dependency versions in a buildout.  This would be measured in part by willingness to do or fund the work required.
This discussion should happen on the distutils list, rather than in a github issue.

**Comment by freddrake** [comment]
While there would be some work to switch from eggs to wheels, or to support both, that doesn't seem particularly significant.  We should be able to install wheels in much the same way as eggs, ending up with a directory of separate installations of any number of versions of packages.

The real issues I see are:

1. the (non) availability of an API to deal with the sdist --> built package --> installed package operations, and

2. the need for something like entry points for packages installed from wheels.

(It's possible someone's done something about the later, and I've just missed it.)

**Comment by guyzmo** [comment]
I'm reaching this issue as I also started to think that buildout needs a bit of refreshment. The whole python community is going towards usage of virtualenv+wheels, and though I agree it got a bunch of advantages, I still find buildout to be neater and more flexible as some sort of development package.

So I'd be happy to see buildout integrate nicely with virtualenv (taking advantage of requirements.txt, and using pip to install packages). Actually, to me "integrate nicely" means the two following use cases:

- when a project has been configured with `buildout.cfg`, still make it possible for people using virtualenv to deploy an environment without having a clue about buildout
- when a project has been configured for virtualenv, make it possible to just run `buildout` and have a `buildout.cfg` generated and stuff installed like it was a buildout project.

**Comment by enkidulan** [comment]
I really would like to not have have any changes in buildout just for making it more compatible with virtualenv and that awful way of assembling envs via "pip install requirements.txt + some scripts to do the rest".

@guyzmo , about 
> when a project has been configured for virtualenv, make it possible to just run buildout and have a buildout.cfg generated and stuff installed like it was a buildout project.

There is number of packages that allows you to create templates like that, kind of old but good https://github.com/srid/modern-package-template, also there is a number of projects boilerplates based on buildout, I use my own for most of project - https://github.com/enkidulan/pyramid_boilerplate

**Comment by jimfulton** [comment]
Someone I respect recently suggested to me that buildout should be more language neutral, and that assembling a Python script should be the responsibility of a recipe with very little support or involvement from buildout.  (In this model, perhaps buildout would use setuptools solely for installing and loading recipes) I think this idea has a lot of merit.  I think it also has a lot of overlap with ideas expressed here. As a bonus, it should be easy to prototype without changes to buildout.

To this end, it would be interesting for someone to prototype this idea by creating a recipe to build Python apps using virtualenv and pip (or whatever).

**Comment by leorochael** [comment]
On the surface, buildout already looks like it's language neural. The fact that the recipes of parts can basically do whatever they want, but the parts get executed in the order of the dependency graph that is traced by the variables shared between them, is a very powerful notion that is not married to any one specific language.

What belies the language neutrality of buildout is the intimate relationship between buildout proper and `zc.recipe.egg` (and derived recipes).

When people say that buildout needs to support wheels, this can mean one of two things:

 * buildout and it's recipes and extensions should be installable from wheels, and be able to have wheel-only packages as dependencies.
 * `zc.recipe.egg` should be able to install wheels, or packages with wheel-only dependencies.

While the former should be possible, people usually actually mean the latter.

When they say that they want buildout to work more like pip, or to be able to work with `requirements.txt`, what they actually want is something like `zc.recipe.egg` to do those things.

But the relationship between buildout and `zc.recipe.egg` is rather incestuous:

 * Some settings from `[buildout]` affect the result of all `zc.recipe.egg` parts the exact same way that they affect the recipe and extension packages of buildout itself. For example `buildout:develop` and `buildout:versions` (my pet peeve, `zc.recipe.egg` should have a per part `versions`).
 * `buildout init` accepts arguments that creates a `zc.recipe.egg` part by default.
 * There used to be a `python` setting in `zc.recipe.egg`, to select a different python interpreter for a part, but it went away, and for perfectly good reasons if you assume that buildout anc `zc.recipe.egg` are just different faces of the same code.
 * The buildout documentation emphasizes `zc.recipe.egg` a lot

All this is to say that I'm +1 on the idea that buildout should be language neutral (with recipes and extensions still written in Python, of course), and that, as @jimfulton mentioned, all that is required from a technical point of view is for someone to write a `zc.recipe.egg` alternative (or alternatives for other languages) that do not rely on the internal state of Python packages for buildout proper, which should only affect itself, it's extensions and its recipe packages.

Buildout would remain an arbitrer of "who installed what where" and should maybe even start tracking "who put which script in `./bin`" (and maybe track other files in other places), with information provided from the recipes, so that parts can avoid stepping onto each other's toes.

Then, someone needs to go over the buildout documentation and deemphasize 'zc.recipe.egg` and related packages.

**Comment by sirex** [comment]
Yesterday I decided to experiment by switching from `zc.recipe.egg` to `pip` and `pip-tools` https://github.com/sirex/manopozicija.lt/commit/5a46e4bec10c3dbd1c1721ba702ae4f254323548 .

So now I use buildout for:

- downloading web assets using `hexagonit.recipe.download`
- generating `bin/django` script with `djangorecipe`
- sharing buildout settings with Django via `exportrecipe`
- building tags file with `z3c.recipe.tag`

Buildout is no longer responsible for installing Python packages, almost all Python packages are installed via `pip` and `requirements.txt` file is managed using `pip-tools`. To make this work, I moved all project Python dependencies from `setup.py` to `requirements.in`.

Buildout still installs recipe packages with their dependencies in his old ways.

**Comment by guyzmo** [comment]
> >  when a project has been configured for virtualenv, make it possible to just run buildout and have a buildout.cfg generated and stuff installed like it was a buildout project.
> There is number of packages that allows you to create templates like that, kind of old but good https://github.com/srid/modern-package-template, also there is a number of projects boilerplates based on buildout, I use my own for most of project - https://github.com/enkidulan/pyramid_boilerplate

I'm actually quite unhappy with "boilerplates". I've been using copy/pasting of my own `buildout.cfg` for all my projects, but it's a pain when I clone someone else's project that assumes that you'll be happy using virtualenv (or any \*env alternative) and have to craft a `buildout.cfg` for that project (or worst: use virtualenv, which despites all my efforts, I still don't like).

So today, I decided that was enough, that I had to code some tool to make it easier to keep on working with buildout in a pip/virtualenv world: [guyzmo/buildstrap](https://github.com/guyzmo/buildstrap). Basically it's a tool to generate a buildout.cfg based on requirements files (taking advantage of the `collective.recipe.pip` recipe to use the repository's requirements.txt list of packages).

that tool is stupid simple, just run at the root of your project:

    % buildstrap run package_name requirements.txt

and it'll generate the following `buildout.cfg`:

```
[buildout]
newest = false
parts = httpie
develop = .
eggs-directory = ${buildout:directory}/var/eggs
develop-eggs-directory = ${buildout:directory}/var/develop-eggs
parts-directory = ${buildout:directory}/var/parts

[package_name]
recipe = zc.recipe.egg
eggs = ${package_name-requirements.txt:eggs}
        package_name

[package_name-requirements.txt]
recipe = collective.recipe.pip
configs = requirements.txt
```

if you want more parts, you can do:

    buildstrap project requirements.txt test=requirements-test.txt doc=requirements-doc.txt

which will generate the `project`, `test` and `doc` parts.
